Add android-key attestation format support

jbpin opened the pull request:
Add a new attestationFormat for ACME device-attest-01 challenge to support Android attestation (android-key) as defined by WebAuthn and modify by RFC (sig use key authorization).

The implementation involve adding a CRL. ACME provider support a new configuration key called RootCRLs (rootCRLs in json). When 'android-key' is specified in attestationFormat and the list is not provided by the configuration, the list will be populated and updated
automatically based on the validation implementation procedure. Other ACME challenge could use IsRootRevoked and RootCRLs in the future independantly to android-key or device-attest-01 challenge.

#### Name of feature: 
android-key attestation support for device-attest-01 challenge

#### Pain or issue this feature alleviates: 
All to use device identifier certificate issuing on Android devices

#### Why is this important to the project (if not answered above): 
The project already support Apple, so supporting Android is a plus. 

#### Is there documentation on how to use this feature? If so, where?
This PR only provide a new attestationFormat called android-key 

#### In what environments or workflows is this feature supported?
ACME provisionner with device_attest_01 and attestationFormat android-key. The device need to be managed by a MDM and an ACME client for android should be deployed

#### In what environments or workflows is this feature explicitly NOT supported (if any)?

#### Supporting links/other PRs/issues:

💔Thank you!

jbpin commented:
Hi, 
This PR can be test with a provisioner config like this : 
```
{
            "attestationFormats": [
              "android-key"
            ],
            "challenges": [
              "device-attest-01"
            ],
            "claims": {
              "defaultTLSCertDuration": "2h",
              "maxTLSCertDuration": "8h"
            },
            "name": "android-acme",
            "type": "ACME"
          },
```

However, this PR will require an update to linkedca for two reason: 
- `android-key` is not supported on attestionFormat [in the proto].(https://github.com/smallstep/linkedca/blob/339b487a925e0cf176a55133a755a1a40e602c35/spec/linkedca/provisioners.proto#L149).
- the new RootCRLs can't be included in ether ca.json nor step cli, so we can't override the default behavior that consist of loading the Android CRL Json

The CRL List is cached for 24h and refreshed when android-key attestation validation occurred passed this delay. 

I will be happy to reply to any comments or suggestions.

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+		break
+	}
+
+	// validate challenge
+	if string(data.Attestation.AttestationChallenge) != keyAuth {
+		return nil, NewDetailedError(ErrorBadAttestationStatementType, "Challenge mismatach: "+string(data.Attestation.AttestationChallenge))
+	}
```
```suggestion
	// validate challenge
	if string(data.Attestation.AttestationChallenge) != keyAuth {
		return nil, NewDetailedError(ErrorBadAttestationStatementType, fmt.Sprintf("challenge mismatch; expected %q, got %q", keyAuth, string(data.Attestation.AttestationChallenge))
	}
```

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+	}
+
+	// validate challenge
+	if string(data.Attestation.AttestationChallenge) != keyAuth {
```
Consider using `subtle.ConstantTimeCompare` to compare these.

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+			return nil, NewDetailedError(ErrorBadAttestationStatementType, "Root certificate not signed by Android")
+		}
+	default:
+		return nil, NewDetailedError(ErrorBadAttestationStatementType, "Invalid root certificate signature algorithm")
```
```suggestion
		return nil, NewDetailedError(ErrorBadAttestationStatementType, "invalid root certificate key type")
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+	for k := range CRLResponse.Entries {
+		keys = append(keys, k)
+	}
+	p.RootCRLs = keys
```
This field should be renamed. It contains the revoked keys themselves; not multiple CRLs. Maybe `p.RevokedAndroidAttestationKeySerials`.

hslatman replied:
Addressed in #2749 by refactoring the CRL check.

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -120,6 +128,8 @@ type ACME struct {
 	AttestationRoots    []byte   `json:"attestationRoots,omitempty"`
 	Claims              *Claims  `json:"claims,omitempty"`
 	Options             *Options `json:"options,omitempty"`
+	RootCRLs            []string `json:"rootCRLs,omitempty"`
```
It would be preferable to not have to define `RootCRLs` on the provisioner. The existing properties are all meant as configuration, whereas the `RootCRLs` is currently used as state. 

If we don't expect users to have to set a different key revocation status URL, then `RootCRLs` (also see the other comment to rename it) can be unexported, and just be handled as state. Maybe all Android revocation state can be contained in a new struct as further improvement.

jbpin replied:
I think this is based on the comment you made on discord: We may also want to be able to provide revoked keys in a way that doesn't rely on the Google endpoint for those same purposes.

I've certainly misunderstand your point so I made it as a configuration. Can be removed.

hslatman replied:
Addressed in #2749 by refactoring the CRL configuration.

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+// fetch CRL https://android.googleapis.com/attestation/status and build a list of serial number
+func (p *ACME) initializeAndroidCRL() error {
+	log.Printf("Updating Android CRL list for %s ACME provisioner", p.Name)
+	var CRLResponse struct {
```
```suggestion
	var crlResponse struct {
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+		return fmt.Errorf("unexpected Android CRL response %d: %s", res.StatusCode, string(bodyBytes))
+	}
+
+	if err := json.NewDecoder(res.Body).Decode(&CRLResponse); err != nil {
```
```suggestion
	if err := json.NewDecoder(res.Body).Decode(&crlResponse); err != nil {
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+	}
+
+	// Extract keys into a slice
+	keys := make([]string, 0, len(CRLResponse.Entries))
```
```suggestion
	keys := make([]string, 0, len(crlResponse.Entries))
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+
+	// Extract keys into a slice
+	keys := make([]string, 0, len(CRLResponse.Entries))
+	for k := range CRLResponse.Entries {
```
```suggestion
	for k := range crlResponse.Entries {
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
 
+const ANDROID_ATTESTATION_STATUS_URL = "https://android.googleapis.com/attestation/status"
+
+// fetch CRL https://android.googleapis.com/attestation/status and build a list of serial number
+func (p *ACME) initializeAndroidCRL() error {
```
```suggestion
// fetch CRL https://android.googleapis.com/attestation/status and build a list of revoked serial numbers
func (p *ACME) fetchAndroidCRL() error {
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -392,3 +442,12 @@ func (p *ACME) IsAttestationFormatEnabled(_ context.Context, format ACMEAttestat
+// It will also be in charge of updating the list periodically if no CRL list is provided at configuration.
+func (p *ACME) IsRootRevoked(serialNumber string) bool {
+	if slices.Contains(p.AttestationFormats, "android-key") && !p.androidCRLTimeout.IsZero() && time.Now().After(p.androidCRLTimeout) {
+		p.initializeAndroidCRL()
```
I kinda like how this is done lazily, but I think it would be better if it were performed on a timer and taking into account caching data on the Google side. It might be sufficient initially, though.
```suggestion
		p.fetchAndroidCRL()
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -392,3 +442,12 @@ func (p *ACME) IsAttestationFormatEnabled(_ context.Context, format ACMEAttestat
+// IsRootRevoked return a true if the serialNumber is part of the list
+// It will also be in charge of updating the list periodically if no CRL list is provided at configuration.
+func (p *ACME) IsRootRevoked(serialNumber string) bool {
+	if slices.Contains(p.AttestationFormats, "android-key") && !p.androidCRLTimeout.IsZero() && time.Now().After(p.androidCRLTimeout) {
```
Consider an additional early return for clarity

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -392,3 +442,12 @@ func (p *ACME) IsAttestationFormatEnabled(_ context.Context, format ACMEAttestat
+	if slices.Contains(p.AttestationFormats, "android-key") && !p.androidCRLTimeout.IsZero() && time.Now().After(p.androidCRLTimeout) {
+		p.initializeAndroidCRL()
+	}
+	return len(p.RootCRLs) > 0 && slices.Contains(p.RootCRLs, serialNumber)
```
It's sufficient to just check using `Contains`, as it's backed by `range`:
```suggestion
	return slices.Contains(p.RootCRLs, serialNumber)
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -392,3 +442,12 @@ func (p *ACME) IsAttestationFormatEnabled(_ context.Context, format ACMEAttestat
 	return p.attestationRootPool, p.attestationRootPool != nil
 }
+
+// IsRootRevoked return a true if the serialNumber is part of the list
+// It will also be in charge of updating the list periodically if no CRL list is provided at configuration.
+func (p *ACME) IsRootRevoked(serialNumber string) bool {
```
Also see comment w.r.t. fetching the serials. It would be better to extract that out of this function, to make it responsible for a single thing.
```suggestion
// IsRootRevoked returns true if the provided serialNumber is in the list of revoked 
// certificate serial number.
// It will also be in charge of updating the list periodically if no CRL list is provided at configuration.
func (p *ACME) IsRootRevoked(serialNumber string) bool {
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+			Reason string `json:"reason"`
+		} `json:"entries"`
+	}
+	res, err := http.Get(ANDROID_ATTESTATION_STATUS_URL)
+	if err != nil {
+		return fmt.Errorf("client: error making Android CRL request: %s\n", err)
+	}
```
Can you change this into using the `http.Client` that's available on the `Controller`?

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
+			Reason string `json:"reason"`
+		} `json:"entries"`
+	}
+	res, err := http.Get(ANDROID_ATTESTATION_STATUS_URL)
```
```suggestion
	res, err := http.Get(androidAttestationStatusURL)
```

hslatman commented on authority/provisioner/acme.go:
```diff
@@ -216,10 +226,50 @@ func (p *ACME) Init(config Config) (err error) {
 	}
 
+	if slices.Contains(p.AttestationFormats, "android-key") && len(p.RootCRLs) == 0 {
+		p.initializeAndroidCRL()
```
```suggestion
		p.fetchAndroidCRL()
```

hslatman reviewed (commented):
Hey @jbpin,

Looking quite good overall. Many suggestions are stylistic; feel free to combine them into your own commit(s). 

I'd like to test this myself with a device too to inspect the full X5C that's provided. Or, alternatively, it would be nice to get a chain from your testing. There's some surprising thing in the code and the Google docs, so I just want to make sure things are correct.

Functionally, and kind of expected, the biggest thing that needs some attention/changes is fetching and refreshing the revoked attestation certificate serials. Eventually that logic needs to become "more pluggable" because of the way we integrate changes into our hosted version. Concretely, it would not be necessary to refetch the revocation list for each hosted authority, as it would simply be retrieving the same list for each. I think it'll come down to going with a basic version for now, and then we may do some changes to make it compatible with our platform ourselves, but we'll have to see how things roll.

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
 	return data, nil
 }
 
+// Android Root CA
+// https://developer.android.com/privacy-and-security/security-key-attestation#root_certificate
+const AndroidRootCAPubKey = `-----BEGIN PUBLIC KEY-----
+MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAr7bHgiuxpwHsK7Qui8xU
```
Also see e.g. the `apple` format, which relies on the default hardcoded Apple attestation root CA, but allows overriding it by setting the existing `attestationRoots` configuration on the provisioner level.

hslatman replied:
Addressed in #2749.

jbpin commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+// https://source.android.com/docs/security/features/keystore/attestation#id-attestation
+var oidAndroidAttestation = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 1, 17}
+
+type androidKeyAttestationData struct {
+	Certificate *x509.Certificate
+	Fingerprint string
+	Attestation *attestation.KeyDescription
+}
```
I imagine it is possible. The difference between Android Attestation and Apple Attestation, if I understand it, is that in Apple each data is a separate OID when Android is a structure in one OID. Parsing this value to the attestation KeyDescription object coming from the lib introduce in this implementation. We may still extract value necessary then discard the whole object.

jbpin commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+		}
+	}
+
+	if root == nil {
+		return nil, NewDetailedError(ErrorBadAttestationStatementType, "missing root certificate in x5c chain")
+	}
```
Yes the root is part of the chain. 

> check that the attestation certificate chain contains a root certificate that is signed with the Google attestation root key and that the attestationSecurityLevel element within the [key description](https://source.android.com/docs/security/features/keystore/attestation#keydescription-fields) data structure is set to the TrustedEnvironment security level or to the StrongBox security level.

jbpin committed:
Update authority/provisioner/acme.go

Co-authored-by: Herman Slatman <hslatman@users.noreply.github.com>

jbpin committed:
Apply suggestions from code review

Co-authored-by: Herman Slatman <hslatman@users.noreply.github.com>

jbpin committed:
switch compare to subtle

jbpin committed:
impl. suggestin from MR

jbpin committed:
revert findAndroidAttestationCert to follow guideline

jbpin committed:
Add android-key attestation format support

Add a new attestationFormat for ACME device-attest-01 challenge to
support Android attestation (android-key) as defined by WebAuthn and
modify by RFC (sig use key authorization).

The implementation involve adding a CRL. ACME provider support a new
configuration key called RootCRLs (rootCRLs in json).
When 'android-key' is specified in attestationFormat and the list is
not provided by the configuration, the list will be populated and
updated
automatically based on the validation implementation procedure.
Other ACME challenge could use IsRootRevoked and RootCRLs in the future
independantly to android-key or device-attest-01 challenge.

jbpin committed:
fix: android root CA error

jbpin committed:
implement certificat based validation for root certificate in android
attestation

jbpin committed:
apply suggestion from merge request

jbpin committed:
apply suggestion to MR

jbpin committed:
apply suggestion and refactor from MR

jbpin committed:
Merge branch 'master' into android

hslatman commented on acme/challenge.go:
```diff
@@ -36,6 +37,7 @@ import (
 	"go.step.sm/crypto/x509util"
 	"golang.org/x/exp/slices"
 
+	"github.com/mbreban/attestation"
```
We've forked it under `github.com/smallstep/android-attestation`.

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+		return nil, err
+	}
+
+	// Parse attestation data:
+	// find the attestation certificate
+	attCert, err := findAndroidAttestationCert(certs)
+	if err != nil {
```
Addressed in #2749 by iterating from the end (root) of the chain to the start (leaf).

hslatman commented on acme/challenge.go:
```diff
@@ -1367,6 +1399,195 @@ func doAppleAttestationFormat(_ context.Context, prov Provisioner, _ *Challenge,
+			return nil, NewDetailedError(ErrorBadAttestationStatementType, "Root certificate not signed by Android")
+		}
+	default:
+		return nil, NewDetailedError(ErrorBadAttestationStatementType, "Invalid root certificate signature algorithm")
```
Covered in #2749.

hslatman commented:
Closing this, as its functionality has been merged via #2749.
